src/self_driving/pilot_util.cpp

```diff
@@ -0,0 +1,150 @@
+
+  // Commented out since currently not performing any actions on aggregated data
+  auto aggregated_data = reinterpret_cast<metrics::PipelineMetricRawData *>(
+      metrics_manager->AggregatedMetrics()
+          .at(static_cast<uint8_t>(metrics::MetricsComponent::EXECUTION_PIPELINE))
+          .get());
+  //  NOISEPAGE_ASSERT(aggregated_data->pipeline_data_.size() >= forecast->query_id_to_params_.size(),
+  //                   "Expect at least one pipeline_metrics record for each query");
+  SELFDRIVING_LOG_INFO("Printing qid and pipeline id to sanity check pipeline metrics recorded");
+  for (auto it = aggregated_data->pipeline_data_.begin(); it != aggregated_data->pipeline_data_.end(); it++) {
+    SELFDRIVING_LOG_INFO(
+        fmt::format("qid: {}; ppl_id: {}", static_cast<uint>(it->query_id_), static_cast<uint32_t>(it->pipeline_id_)));
+  }
+
+  return aggregated_data->pipeline_data_;
+}
+
+void PilotUtil::InferenceWithFeatures(
+    common::ManagedPointer<modelserver::ModelServerManager> ms_manager,
+    const std::list<metrics::PipelineMetricRawData::PipelineData> &pipeline_data,
+    std::list<std::tuple<execution::query_id_t, execution::pipeline_id_t, std::vector<std::vector<double>>>>
+        *pipeline_to_prediction) {
+  std::unordered_map<ExecutionOperatingUnitType, std::vector<std::vector<double>>> ou_to_features;
+  std::list<std::tuple<execution::query_id_t, execution::pipeline_id_t,
+                       std::vector<std::tuple<ExecutionOperatingUnitType, uint64_t>>>>
+      pipeline_to_ou_position;
+
+  PilotUtil::GroupFeaturesByOU(&pipeline_to_ou_position, pipeline_data, &ou_to_features);
+  NOISEPAGE_ASSERT(ms_manager->ModelServerStarted(), "Model Server should have been started");
+  std::string project_build_path = getenv(Pilot::BUILD_ABS_PATH);
+  std::unordered_map<ExecutionOperatingUnitType, std::vector<std::vector<double>>> inference_result;
+  for (auto &ou_map_it : ou_to_features) {
+    auto res =
+        ms_manager->DoInference(selfdriving::OperatingUnitUtil::ExecutionOperatingUnitTypeToString(ou_map_it.first),
+                                project_build_path + Pilot::SAVE_PATH, ou_map_it.second);
+    if (!res.second) {
+      throw PILOT_EXCEPTION("Inference through model server manager has error", common::ErrorCode::ERRCODE_WARNING);
+    }
+    inference_result.emplace(ou_map_it.first, res.first);
+  }
+
+  // PilotUtil::ParseInferenceResult(pipeline_to_ou_position, inference_result, pipeline_to_prediction);
+}
+
+void PilotUtil::GroupFeaturesByOU(
+    std::list<std::tuple<execution::query_id_t, execution::pipeline_id_t,
+                         std::vector<std::tuple<ExecutionOperatingUnitType, uint64_t>>>> *pipeline_to_ou_position,
+    const std::list<metrics::PipelineMetricRawData::PipelineData> &pipeline_data,
+    std::unordered_map<ExecutionOperatingUnitType, std::vector<std::vector<double>>> *ou_to_features) {
+  std::vector<std::tuple<ExecutionOperatingUnitType, uint64_t>> ou_positions;
```

Comment:
nit: We usually use `std::pair` instead of `std::tuple` when there're exactly two values.